flask_sqlalchemy/__init__.py

```diff
@@ -730,19 +730,23 @@ class User(db.Model):
        naming conventions among other, non-trivial things.
     """
 
-    def __init__(self, app=None, use_native_unicode=True, session_options=None, metadata=None):
+    def __init__(self, app=None, use_native_unicode=True, session_options=None,
+                 metadata=None, query_class=BaseQuery, model_class=Model):
 
         if session_options is None:
             session_options = {}
 
         session_options.setdefault('scopefunc', connection_stack.__ident_func__)
+        session_options.setdefault('query_cls', query_class)
         self.use_native_unicode = use_native_unicode
         self.session = self.create_scoped_session(session_options)
-        self.Model = self.make_declarative_base(metadata)
-        self.Query = BaseQuery
+        self.Model = self.make_declarative_base(model_class, metadata)
+        # ugly hack, is there a better way?
+        self.Model.query_class = query_class
```

Comment:
I added a check to respect the existing `query_class` attribute on models, and this potential replacement was moved into `SQLAlchemy.make_declarative_base`. However, I'm wondering if this'll cause confusion with some people. Providing a model base with your desired `query_class` will work though.